src/cpp/src/visual_language/qwen2vl/classes.cpp

```diff
@@ -588,19 +748,33 @@ EncodedImage VisionEncoderQwen2VL::encode_with_imagepreprocess_cpp(const ov::Ten
     encoder.infer();
 
     const ov::Tensor& infer_output = encoder.get_output_tensor();
-    ov::Tensor image_features(infer_output.get_element_type(), infer_output.get_shape());
-    std::memcpy(image_features.data(), infer_output.data(), infer_output.get_byte_size());
-    ImageSize resized_source_size{grid_h, grid_w};
-    return {std::move(image_features), resized_source_size};
+    // Just avoid to multiple copy.
+    if (frame_id == 0u) {
+        auto out_shape = infer_output.get_shape();
+        out_shape[0] = out_shape[0] * frame_num;
+        out_tensor = ov::Tensor(infer_output.get_element_type(), out_shape);
+    }
+
+    std::memcpy(reinterpret_cast<uint8_t*>(out_tensor.data()) + frame_id * infer_output.get_byte_size(),
+                infer_output.data(),
+                infer_output.get_byte_size());
+    out_rsz_size = ImageSize{grid_h, grid_w};
 }
 
 // keep both implementations for comparison and testing, here is the ov version
-EncodedImage VisionEncoderQwen2VL::encode_with_imagepreprocess_ov(const ov::Tensor& image, const ov::AnyMap& config_map) {
+// input multiple images, process based on video.
+// input single image, process based on image.
+void VisionEncoderQwen2VL::encode_with_imagepreprocess_ov(const std::vector<ov::Tensor>& images,
+                                                          const ov::AnyMap& config_map,
+                                                          ov::Tensor& out_tensor,
+                                                          ImageSize& out_rsz_size,
+                                                          size_t frame_num,
+                                                          size_t frame_id) {
     CircularBufferQueueElementGuard<ov::InferRequest> infer_request_guard(this->m_ireq_queue_vision_encoder.get());
     ov::InferRequest& encoder = infer_request_guard.get();
     ProcessorConfig config = utils::from_any_map(config_map, m_processor_config);
 
-    ov::Shape image_shape = image.get_shape();
+    ov::Shape image_shape = images[0].get_shape();
```

Comment:
Are we sure that all images will be of the same shape?
What will be if they are of different shapes?

Reply:
Sure, `encode_with_imagepreprocess_ov`, only input one image when processing image.
when images.size > 1, it means it is process video case. we default think all video frames have same size.

here is my comments
https://github.com/xipingyan/openvino.genai/blob/fb8c1f3c52afd96ec9a9d10a0a446392d9570c8d/src/cpp/src/visual_language/qwen2vl/classes.cpp#L753-L756